validphys2/src/validphys/commondataparser.py

```diff
@@ -40,21 +108,805 @@
+    data_uncertainties: list[ValidPath]
+
+    # The central data is optional _only_ for
+    # positivity datasets, and will be checked as soon as the class is instantiated
+    data_central: Optional[ValidPath] = None
+
+    # Optional data
+    theory: Optional[TheoryMeta] = None
+    tables: Optional[list] = dataclasses.field(default_factory=list)
+    npoints: Optional[list] = dataclasses.field(default_factory=list)
+    variants: Optional[ValidVariants] = dataclasses.field(default_factory=dict)
+    applied_variant: Optional[str] = None
+    ported_from: Optional[str] = None
+
+    # Derived quantities:
+    # Note that an observable without a parent will fail in many different ways
+    _parent: Optional[Any] = None
+
+    def __post_init__(self):
+        """
+        Small modifications for better compatibility with the rest of validphys
+        """
+        # Since vp will rely on the kinematics being 3 variables,
+        # fill the extra with whatever can be found in the kinematics dictionary
+        # otherwise just fill with extra_x
+        if len(self.kinematic_coverage) < 3:
+            unused = list(set(self.kinematics.variables) - set(self.kinematic_coverage))
+            diff_to_3 = 3 - len(self.kinematic_coverage)
+            if unused:
+                self.kinematic_coverage += unused[diff_to_3:]
+            else:
+                self.kinematic_coverage += [f"extra_{i}" for i in range(diff_to_3)]
+
+        object.__setattr__(self, 'process_type', self.process_type.upper())
+
+    def check(self):
+        """Various checks to apply manually to the observable before it is used anywhere
+        These are not part of the __post_init__ call since they can only happen after the metadata
+        has been read and the observable selected.
+        """
+        # Check that the data_central is empty if and only if the dataset is a positivity set
+        if self.data_central is None and not self.is_lagrange_multiplier:
+            raise ValidationError(f"Missing `data_central` field for {self.name}")
+
+        # Check that plotting.plot_x is being filled
+        if self.plotting.plot_x is None:
+            ermsg = f"No variable selected as x-axis in the plot for {self.name}. Please add plotting::plot_x."
+            if self.plotting.x is not None:
+                ermsg += "If you are using `plotting::x` please change it to `plotting::plot_x`"
+            raise ValidationError(ermsg)
+
+        # Ensure that all variables in the kinematic coverage exist
+        for var in self.kinematic_coverage:
+            if var not in self.kinematics.variables:
+                raise ValidationError(
+                    f"Variable {var} is in `kinematic_coverage` but not included in `kinematics` for {self.name}"
+                )
+
+        if len(self.kinematic_coverage) > 3:
+            raise ValidationError(
+                "Only a maximum of 3 variables can be used for `kinematic_coverage`"
+            )
+
```

Comment:
One thing that I have found myself often when implementing new commondata was the fact that the number of of elements inside the `bins` key of `uncertainties.yaml` do not correspond to the number of datapoints. Would it also be possible to check this here and outputs instructive errors if not.

Reply:
You mean, you'd like an explicit error when that's the case? (I guess right now it fails in some magic way?)

Reply:
In what I observed, it fails only when computing the $\chi^2$ due to a mismatch in shapes but with non-instructive errors making it hard understanding what went wrong. On the other hand, I don't want to open the pandora's box here as one could argue that a test that checks that the number of kinematic points and datapoints also match is required here, and other tests as well.

Reply:
Ok, I'll add the check. This can only be done upon load though (so you can use it wrongly for a while before realizing)